### ISSUE TITLE
Preload Blocks Issue

### ISSUE THREAD

**Issue by CreativeDive** [AdvancedCustomFields/acf#407]
Hey @elliotcondon,

first, I want to say that your preload block feature is awesome and works great for the most part. But I discovered a case with a problem.

A specific block was not loaded after refreshing the editor page. If I deactivate `define('ACF_EXPERIMENTAL_PRELOAD_BLOCKS', true);` this block is loaded every time the page is refreshed.

I found out that this happens when a style tag is inserted right before the block content.

With **`define('ACF_EXPERIMENTAL_PRELOAD_BLOCKS', true);`** the block does not load after refreshing the page.

This is the result:

```
<div id="block-f1a76aca-54de-4ccd-afb7-093b7e456354" tabindex="0" role="group" aria-label="Block: Image / Text" data-block="f1a76aca-54de-4ccd-afb7-093b7e456354" data-type="acf/image-text" data-title="Image / Text" class="block-editor-block-list__block wp-block is-selected">
   <div class="acf-block-component acf-block-body">
      <div>
         <style>#image_text_id_5fa3dfe771699:not(.blank) .item { background-color: #dd685b }</style>
      </div>
   </div>
</div>
```

When I go to the block options and change some settings in the sidebar, the block shows up correctly.

With **`define('ACF_EXPERIMENTAL_PRELOAD_BLOCKS', false);`** the block is loaded correctly every time:

This is the result:

```
<div id="block-575de50b-47fa-4f19-99fb-a576e6616e59" tabindex="0" role="group" aria-label="Block: Image / Text" data-block="575de50b-47fa-4f19-99fb-a576e6616e59" data-type="acf/image-text" data-title="Image / Text" class="block-editor-block-list__block wp-block is-hovered">
   <div class="acf-block-component acf-block-body">
      <div>
         <div class="acf-block-preview">
            <style>#image_text_id_5fa3e0638d744:not(.blank) .item { background-color: #dd685b }</style>
            <div class="block image-text-block left blank wp-block wphave-block has-background-color" id="image_text_id_5fa3e0638d744">
               <div class="items">
                  <div class="item-box">
                     <div class="item element">

HERE COMES THE CONTENT

                     </div>
                  </div>
               </div>
            </div>
         </div>
      </div>
   </div>
</div>
```
Do you have an idea why this block dosen't load with **ACF_EXPERIMENTAL_PRELOAD_BLOCKS** if a style tag is included?

**Comment by elliotcondon** [comment]
Hi @CreativeDive 

Thanks for the bug report - I was starting to get worried with the lack of issues 😅.
Can you send me over some example code to quickly setup the same issue?

**Comment by CreativeDive** [comment]
I my case I get this issue, if I select a custom background color for a specific block. After selecting the background color a style tag was included before the block content. And the result is, this block was not preloaded.

Here is the code of the block:

```
<?php 

/****************
* IMAGE / TEXT - BLOCK
****************/

$block = isset( $block ) ? $block : '';
$id = 'image_text_' . uniqid('id_');

// Get Item Settings
$design = ACF_GF('imgtxt_design_item_design'); // --> Clone field
$padding = ACF_GF('item_spacing');
$appearance = ACF_GF('item_style');
$box_shadow = ACF_GF('item_box_shadow');
$border_radius = ACF_GF('item_border_radius');
$alignment = ACF_GF('imgtxt_image_alignment');

// Get the image
$img_text_image = ACF_GF('imgtxt_image');
$img_text_image = wphave_image( $img_text_image, 'large' );
$img_shape = wphave_acf_select_value( ACF_GF('image_shape') );
$effect = ACF_GF('animated_image_effect'); 
$effectduration = ACF_GF('animated_image_effect_duration'); 

// Get the colum width
$img_column = ACF_GF('imgtxt_column_width');
$content_column = 100 - $img_column;
$left_col = $content_column;
$right_col = $img_column;
if( $alignment === 'left' ) {
	$left_col = $img_column;
	$right_col = $content_column;
}

// Get the colors
$border_color = ACF_GF('imgtxt_border_color');
$bg_color = ACF_GF('imgtxt_bg_color');
$shadow_color = ACF_GF('imgtxt_shadow_color');

// Get block wrapper args
$args = array(
	'selector' => 'image-text-block',
	'design' => $design,
	'appearance' => array(
		'padding' => $padding,
		'shape' => $appearance,
		'shadow' => $box_shadow,
		'shadow_color' => $shadow_color,
		'border_radius' => $border_radius,
		'border_color' => $border_color,
		'bg_color' => $bg_color,
	),
	'alignment' => $alignment,
	'block' => $block,
	'id' => $id,
);

/****************
* OUTPUT
****************/
	
do_action( 'block_wrapper_before', $args ); ?>

<div class="item-box">
	<div class="item element">

		<div class="item-wrapper" style="grid-template-columns: <?php echo esc_html( $left_col ); ?>% <?php echo esc_html( $right_col ); ?>%;">

			<div class="item-img element">

				<?php 

				/****************
				* IMAGE
				****************/

				if( $img_text_image ) { ?>
					<div class="<?php if( $effect ) { ?>wow <?php echo esc_html( $effect ); ?><?php } ?>"<?php if( $effectduration ) { ?> data-wow-duration="<?php echo esc_attr( $effectduration ); ?>s"<?php } ?>>

						<?php if( $img_shape !== 'original' ) { ?>

							<div class="image-shape-container <?php echo esc_html( $img_shape ); ?>">
								<div class="img-objectfit">
									<?php echo wp_kses_post( $img_text_image );
									echo wphave_clip_path_shape( $img_shape ); ?>
								</div>
							</div>

						<?php } else {

							echo wp_kses_post( $img_text_image );

						} ?>

					</div>
				<?php } else {
					
					wphave_placeholder_image();
					
				} ?>

			</div>

			<div class="item-content">

				<?php

				/****************
				* ACF INNER BLOCKS - START
				****************/

				// Allow only the following block to include in the header block
				$allowed_blocks = array( 
					// Core blocks
					'core/heading',
					'core/paragraph',
					'core/columns',
					'core/image', 
					'core/list',
					'core/html',
					'core/buttons',
					'core/spacer',
					'core/search',

					// ACF blocks
					'acf/accordion-menu',
					'acf/button',
					'acf/links',
					'acf/content-boxes',
					'acf/coupon',
					'acf/custom-html',
					'acf/custom-teaser',
					'acf/heading',
					'acf/heading-dual',
					'acf/heading-masked',
					'acf/icon-box',
					'acf/icon-list',
					'acf/media-box',
					'acf/separator',
					'acf/site-logo',
					'acf/social-profiles',
					'acf/social-share',
					'acf/tab-menu',
					'acf/toggle-menu',
				);

				?>

				<InnerBlocks allowedBlocks="<?php echo esc_attr( wp_json_encode( $allowed_blocks ) ); ?>" />

				<?php

				/****************
				* ACF INNER BLOCKS - END
				****************/

				?>

			</div>

		</div>

	</div>
</div>

<?php do_action( 'block_wrapper_after', $args );
```

Here is the action which generate the style tag before the block content.

```
add_action( 'block_wrapper_before', function( $args = '' ) {

	$selector = isset( $args['selector'] ) ? $args['selector'] : 'undefined';
	$column = isset( $args['column'] ) ? $args['column'] : '';
	$view = isset( $args['view'] ) ? $args['view'] : '';
	$design = isset( $args['design'] ) ? $args['design'] : 'blank';
	$appearance = isset( $args['appearance'] ) ? $args['appearance'] : '';
	$alignment = isset( $args['alignment'] ) ? $args['alignment'] : '';
	$block = isset( $args['block'] ) ? $args['block'] : '';
	$carousel = isset( $args['carousel'] ) ? $args['carousel'] : '';
	$id = isset( $args['id'] ) ? $args['id'] : '';
	$html_before = isset( $args['html_before'] ) ? $args['html_before'] : '';
	$class = isset( $args['class'] ) ? $args['class'] : ''; 
	$style = isset( $args['style'] ) ? $args['style'] : ''; 

	$classes = array();

	if( $selector ) {
		$classes[] = $selector; 
	}

	if( $class ) {
		if( is_array( $class ) ) {
			$classes[] = implode(' ', $class );
		} else {
			$classes[] = $class;
		}
	}

	if( $alignment ) {
		$classes[] = $alignment; 
	}

	if( isset( $column['column_count'] ) ) {
		$classes[] = 'column-' . $column['column_count']; 
	}

	if( $design ) {
		$classes[] = $design; 
	}

	$border_radius = '';
	$padding = '';
	$border_color = '';
	$bg_color = '';
	$has_shadow = '';
	$shadow_color = '';

	if( $appearance ) {
		if( isset( $appearance['shape'] ) && $appearance['shape'] === 'rounded' ) {
			$classes[] = 'rounded';
		}
		if( isset( $appearance['shadow'] ) && $appearance['shadow'] === 'has_shadow' ) {
			$classes[] = 'has-shadow';
			$has_shadow = true;
		}
		if( isset( $appearance['border_radius'] ) && $appearance['border_radius'] != 0 ) {
			$border_radius = $appearance['border_radius'];
		}
		if( isset( $appearance['padding'] ) && $appearance['padding'] != 'default' ) {
			$padding = $appearance['padding'];
			if( $padding === 'size_0' ) {
				$padding = 'null';
			}
		}
		if( isset( $appearance['border_color'] ) ) {
			$border_color = $appearance['border_color'];
		}
		if( isset( $appearance['bg_color'] ) ) {
			$bg_color = $appearance['bg_color'];
		}
		if( isset( $appearance['shadow_color'] ) ) {
			$shadow_color = $appearance['shadow_color'];
		}
	}

	if( $block ) {
		$classes[] = wphave_block_editor_class( $block ); 
	}

	if( $carousel ) {
		$classes[] = 'is-carousel'; 
	}

	if( $view && ! $carousel ) {
		$classes[] = 'is-grid'; 
	}
	
	if( $bg_color ) {
		$classes[] = 'has-background-color'; 
	}

	$styles = array();

	if( $style ) {
		if( is_array( $style ) ) {
			$styles[] = implode('; ', $style );
		} else {
			$styles[] = $style;
		}
	} 

	/****************
	* CUSTOM INLINE STYLES
	****************/		

	if( $border_radius && $border_radius != 0 || $padding || $border_color || $bg_color || $has_shadow && $shadow_color ) { ?>
		<style>
			<?php

			if( $border_radius && $border_radius != 0 ) { ?>
				#<?php echo esc_html( $id ); ?> .item { border-radius: <?php echo esc_html( $border_radius ); ?>px; overflow: hidden; }
			<?php } 

			if( $padding && $design != 'blank' ) { 
				if( $padding === 'null' ) { ?>
					#<?php echo esc_html( $id ); ?> .item { padding: 0rem; }
				<?php } else { ?>
					#<?php echo esc_html( $id ); ?> .item { padding: <?php echo esc_html( $padding ); ?>rem; }
				<?php }
			}

			if( $border_color ) { ?>
				#<?php echo esc_html( $id ); ?>.bordered .item { border-color: <?php echo esc_html( $border_color ); ?> }
			<?php }

			if( $bg_color ) { ?>
				#<?php echo esc_html( $id ); ?>:not(.blank) .item { background-color: <?php echo esc_html( $bg_color ); ?> }
			<?php }

			if( $has_shadow && $shadow_color ) { ?>
				#<?php echo esc_html( $id ); ?> .item { 
					box-shadow: 0 2px 8px 0 <?php echo esc_html( wphave_hex2rgba( $shadow_color, wphave_opacity('09') ) ); ?>, 0 4px 32px 0 <?php echo esc_html( wphave_hex2rgba( $shadow_color, wphave_opacity('09') ) ); ?>; 
				}
			<?php } ?>
		</style>
	<?php } 
	
	/****************
	* HOOK - WPHAVE [wphave_before_block_template]
	****************/

	do_action('wphave_before_block_template'); 

	/****************
	* WRAPPER START
	****************/

	?>

	<div class="block <?php echo esc_html( implode(' ', $classes ) ); ?>"<?php if( $style ) { ?> style="<?php echo esc_html( implode(' ', $styles ) ); ?>"<?php } ?><?php if( $id ) { ?> id="<?php echo esc_html( $id ); ?><?php } ?>">

		<?php // Insert optional HTML before
		echo wp_kses_post( $html_before ); 
		
			// Add a custom HTML anchor id.
			// --> Can modified on block editor sidebar > "Advanced".
			$html_anchor_id = '';
			if( isset( $block['anchor'] ) && $block['anchor'] ) {
				$html_anchor_id = ' id="' . $block['anchor'] . '"';
			} ?>
		
			<div class="items"<?php echo esc_html( $html_anchor_id ); ?>>

				<?php if( $carousel ) { ?>
					<div class="owl-carousel">
				<?php }

				/* BLOCK CONTENT STARTS HERE */
				/* --> <-- */
				/* BLOCK CONTENT STARTS HERE */

}, 10, 2 );
```

**Comment by elliotcondon** [comment]
Hi @CreativeDive. Thanks for the snippet. I was hoping for a simple demo example that won't require much setup - this looks like a little complicated at first glance. Any chance you can help me out?

**Comment by CreativeDive** [comment]
@elliotcondon for sure, here is a simple examlpe:

```
<?php 

/****************
* IMAGE / TEXT - BLOCK
****************/

$block = isset( $block ) ? $block : '';
$id = 'image_text_' . uniqid('id_');

// Get Item Settings
$design = ACF_GF('imgtxt_design_item_design'); // --> Clone field
$padding = ACF_GF('item_spacing');
$appearance = ACF_GF('item_style');
$box_shadow = ACF_GF('item_box_shadow');
$border_radius = ACF_GF('item_border_radius');
$alignment = ACF_GF('imgtxt_image_alignment');

// Get the image
$img_text_image = ACF_GF('imgtxt_image');
$img_text_image = wphave_image( $img_text_image, 'large' );
$img_shape = wphave_acf_select_value( ACF_GF('image_shape') );
$effect = ACF_GF('animated_image_effect'); 
$effectduration = ACF_GF('animated_image_effect_duration'); 

// Get the colum width
$img_column = ACF_GF('imgtxt_column_width');
$content_column = 100 - $img_column;
$left_col = $content_column;
$right_col = $img_column;
if( $alignment === 'left' ) {
	$left_col = $img_column;
	$right_col = $content_column;
}

$bg_color = ACF_GF('imgtxt_bg_color');

/****************
* CUSTOM INLINE STYLES
****************/		

if( $bg_color ) { ?>
	<style>
		<?php

		if( $bg_color ) { ?>
			#<?php echo esc_html( $id ); ?>:not(.blank) .item { background-color: <?php echo esc_html( $bg_color ); ?> }
		<?php } ?>
	</style>
<?php } 

/****************
* OUTPUT
****************/ ?>
	
<div class="block image-text-block" <?php if( $id ) { ?> id="<?php echo esc_html( $id ); ?><?php } ?>">

	<div class="item-box">
		<div class="item element">

			<div class="item-wrapper" style="grid-template-columns: <?php echo esc_html( $left_col ); ?>% <?php echo esc_html( $right_col ); ?>%;">

				<div class="item-img element">

					<?php 

					/****************
					* IMAGE
					****************/

					if( $img_text_image ) { ?>
						<div class="<?php if( $effect ) { ?>wow <?php echo esc_html( $effect ); ?><?php } ?>"<?php if( $effectduration ) { ?> data-wow-duration="<?php echo esc_attr( $effectduration ); ?>s"<?php } ?>>

							<?php if( $img_shape !== 'original' ) { ?>

								<div class="image-shape-container <?php echo esc_html( $img_shape ); ?>">
									<div class="img-objectfit">
										<?php echo wp_kses_post( $img_text_image );
										echo wphave_clip_path_shape( $img_shape ); ?>
									</div>
								</div>

							<?php } else {

								echo wp_kses_post( $img_text_image );

							} ?>

						</div>
					<?php } else {

						wphave_placeholder_image();

					} ?>

				</div>

				<div class="item-content">

					<?php

					/****************
					* ACF INNER BLOCKS - START
					****************/

					// Allow only the following blocks
					$allowed_blocks = array( 
						// Core blocks
						'core/paragraph',
					);

					?>

					<InnerBlocks allowedBlocks="<?php echo esc_attr( wp_json_encode( $allowed_blocks ) ); ?>" />

					<?php

					/****************
					* ACF INNER BLOCKS - END
					****************/

					?>

				</div>

			</div>

		</div>
	</div>

</div>
```

**Comment by elliotcondon** [comment]
Thanks @CreativeDive. I might setup a even more simple block type without any fields and will let you know what I find.

**Comment by CreativeDive** [comment]
@elliotcondon here is a more simplified block example. The same result. If the block has a custom text color selected, the preload dosen't work.

```
/****************
* MARQUEE - BLOCK
****************/

$block = isset( $block ) ? $block : '';
$id = 'marquee_' . uniqid('id_');

// Get Item Settings
$text = ACF_GF('marquee_text') ?: __('Your text here.', 'wphave');

// Get the colors
$text_color = ACF_GF('marquee_text_color');

/****************
* CUSTOM STYLES
****************/

if( $text_color ) { ?>
	<style>
		<?php

		if( $text_color ) { ?>
			#<?php echo esc_html( $id ); ?> .marquee { color: <?php echo esc_html( $text_color ); ?> }
		<?php } ?>
	</style>
<?php } ?>

<div class="<?php echo $block; ?>">
	<div class="item-box">
		<div class="item element">
			
			<div class="marquee">
				<span>
					<?php echo wp_kses_post( $text ); ?>
				</span>
			</div>
			
		</div>
	</div>
</div>
```

**Comment by grahammtbr** [comment]
Okay, I'm not losing my mind. I have the same issue, I have what's essentially a duplicate of the cover block, all my fields are settings for the <style> section which resides before the block content. My content within the unique ID div is simply <InnerBlocks />.

**Comment by elliotcondon** [comment]
Hi all 👋. I have some good news to share for this issue. 

I'm confident that the issue is due to a missing parent `<div class="acf-block-preview">` element within the preloaded HTML data. Can you please test out a fix by making the following change:

File `/pro/blocks.php` - line 338. From:
```
acf_get_store( 'block-cache' )->set( $attributes['id'], $html );
```

to:
```
acf_get_store( 'block-cache' )->set( $attributes['id'], '<div class="acf-block-preview">' . $html . '</div>' );
```

Please let me know your results 🙌

**Comment by CreativeDive** [comment]
Thanks @elliotcondon 

I've seen you have merged this to ACF 5.9.4.

I can confirm this solve the preloading issue.
